numpy/_core/src/multiarray/stringdtype/casts.c

```diff
@@ -0,0 +1,1510 @@
+            }                                                                  \
+                                                                               \
+            Py_INCREF(given_descrs[0]);                                        \
+            loop_descrs[0] = given_descrs[0];                                  \
+                                                                               \
+            return NPY_##safety##_CASTING;                                     \
+        }
+
+ANY_TO_STRING_RESOLVE_DESCRIPTORS(SAFE)
+ANY_TO_STRING_RESOLVE_DESCRIPTORS(UNSAFE)
+
+
+static NPY_CASTING
+string_to_string_resolve_descriptors(PyObject *NPY_UNUSED(self),
+                                     PyArray_DTypeMeta *NPY_UNUSED(dtypes[2]),
+                                     PyArray_Descr *given_descrs[2],
+                                     PyArray_Descr *loop_descrs[2],
+                                     npy_intp *view_offset)
+{
+    if (given_descrs[1] == NULL) {
+        loop_descrs[1] = stringdtype_finalize_descr(given_descrs[0]);
+    }
+    else {
+        Py_INCREF(given_descrs[1]);
+        loop_descrs[1] = given_descrs[1];
+    }
+
+    Py_INCREF(given_descrs[0]);
+    loop_descrs[0] = given_descrs[0];
+
+    StringDTypeObject *descr0 = (StringDTypeObject *)loop_descrs[0];
+    StringDTypeObject *descr1 = (StringDTypeObject *)loop_descrs[1];
+
+    if ((descr0->na_object != NULL) && (descr1->na_object == NULL)) {
+        // cast from a dtype with an NA to one without, so it's a lossy
+        // unsafe cast
+        return NPY_UNSAFE_CASTING;
+    }
+
+    *view_offset = 0;
+
+    return NPY_NO_CASTING;
+}
+
+static int
+string_to_string(PyArrayMethod_Context *context, char *const data[],
+                 npy_intp const dimensions[], npy_intp const strides[],
+                 NpyAuxData *NPY_UNUSED(auxdata))
+{
+    StringDTypeObject *idescr = (StringDTypeObject *)context->descriptors[0];
+    StringDTypeObject *odescr = (StringDTypeObject *)context->descriptors[1];
+    int in_hasnull = idescr->na_object != NULL;
+    int out_hasnull = odescr->na_object != NULL;
+    const npy_static_string *in_na_name = &idescr->na_name;
+    npy_intp N = dimensions[0];
+    char *in = data[0];
+    char *out = data[1];
+    npy_intp in_stride = strides[0];
+    npy_intp out_stride = strides[1];
+
+    npy_string_allocator *iallocator = NULL;
+    npy_string_allocator *oallocator = NULL;
+    NpyString_acquire_allocator2(idescr, odescr, &iallocator, &oallocator);
+
+    while (N--) {
+        const npy_packed_static_string *s = (npy_packed_static_string *)in;
+        npy_packed_static_string *os = (npy_packed_static_string *)out;
+        if (in != out) {
+            if (in_hasnull && !out_hasnull && NpyString_isnull(s)) {
+                // lossy but this is an unsafe cast so this is OK
+                if (NpyString_pack(odescr->allocator, os, in_na_name->buf,
```

Comment:
```suggestion
                if (NpyString_pack(oallocator, os, in_na_name->buf,
```
No biggy, but would be nice to make sure to use the ones we actually fetched.